### PR TITLE
possible fix for SNAP-1414. passing a reasonable initial size for the…

### PR DESCRIPTION
**Pull request by ahshahid**
… encoder term

I do not fully understand the code , but looks like the initial buffer size of the encoder is determined using columnBatchSize, which we are now passing -1 for sample creation case, which causes index out of bounds. So the change is to use default column batch size to determine defaultBatchTermSize, in case column batch size is -1.

Reviving this PR. As mentioned before, I do not understand the code,  but looks like given an initial size,  it is able to expand, so initial size need not be exact ( which will be the case of sample based buffers which do not have a defined batch size ), so using default column batch size in that case should suffice.
Sumedh/ Vivek pls review the change to bring closure ( one way or the other)

(Fill in the changes here)

## Patch testing

(Fill in the details about how this patch was tested)

## ReleaseNotes.txt changes

(Does this change require an entry in ReleaseNotes.txt? If yes, has it been added to it?)

## Other PRs 

(Does this change require changes in other projects- store, spark, spark-jobserver, aqp? Add the links of PR of the other subprojects that are related to this change)

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnInsertExec.scala [hunk 1]
```diff
@@ -156,7 +157,8 @@ case class ColumnInsertExec(_child: SparkPlan, partitionColumns: Seq[String],
        |  int $defaultRowSize = 0;
        |  ${declarations.mkString("\n")}
        |  $defaultBatchSizeTerm = Math.max(
-       |    (${math.abs(columnBatchSize)} - 8) / $defaultRowSize, 16);
+       |    (${math.abs( if (columnBatchSize > 0) columnBatchSize else Property.ColumnBatchSize.
+            defaultValue.get)} - 8) / $defaultRowSize, 16);
```
Comment (vivekwiz) [inline core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnInsertExec.scala]: Can you please verify from where columnBatchSize is being set to -1?

Two possible places are in ColumnFormatSamplingRelation:
      val buffer = batchCreator.createColumnBatchBuffer(
        // externalTableMetaData.columnBatchSize,
        // Fix for AQP-273, for sample table, the strata needs to remain as a single entry
        -1,

If so, good to change that code.

Comment (vivekwiz) [inline core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnInsertExec.scala]: Had a relook at your comments and code and this is what I think:

Lets pass an additional variable defaultBatchSizeTerm right through ColumnFormatSamplingRelation to ColumnInsertExec.

**Comment by ahshahid** [comment]
I am a bit reluctant passing any additional parameter, as I do not
understand that code thoroughly. My suggestion is that I will checkin the
code with the current change, & let you in conjunction with Sumedh change
it further if you want to. As batch size term is calc. inside based on some
logic, passing it from outside may not look nice.

Regards
Asif

**Comment by vivekwiz** [comment]
You may choose to go ahead with what seems best for you considering all options.

**Pull request merged**